### ISSUE TITLE
TASK: automatic provides for bundled rust crate dependencies

### ISSUE THREAD

**Issue by yselkowitz** [fedora-eln/eln#158]
### What does the ELN SIG need to do?
* Unlike in Fedora, RHEL does not include individual rust crate packages.  Instead, rust-based packages include an additional source tarball with their vendored dependencies (sometimes available upstream, otherwise manually created) with which the package is built.
* Best practices would mandate that these bundled dependencies somehow be recorded in the package.  A few packages in RHEL (9 even) do maintain this manually in the specfile, but others (the majority afaics) do not.
* Manual maintenance of bundled provides is cumbersome and error-prone, and given the increasing use of rust in userspace, is not sustainable.
* The Go SIG recently added means by which bundled provides are automatically generated via `%license vendor/modules.txt` (that file always existing as a result of `go mod vendor`).
* While rust vendoring unfortunately does not have an analogous file, the Fedora rust macros do include macros which parse the dependencies compiled into a "final" rust-based consumer, which some such spec files used to create text files included in `%license`.
* If these macros could somehow be adapted or extended in rust-toolkit, or some other means entirely if needed, to provide the information needed for a new provides filter to parse such a file (either in its current format, a modified one, or a different file entirely) and create the provides therefrom, that would give us a system that could be applied to all rust-based packages within the ELN package set, and then carry over to RHEL.
* As an existing issue in RHEL, this must not block vendoring efforts in ELN, but is IMO something that we want to fix prior to RHEL 10 GA.

**Comment by carlwgeorge** [comment]
> Best practices would mandate that these bundled dependencies somehow be recorded in the package. A few packages in RHEL (9 even) do maintain this manually in the specfile, but others (the majority afaics) do not.

For the RHEL packages that do not, how does Product Security flag these packages for CVEs in the bundled components?  My understanding (which may be flawed or outdated) was that process depends on accurate `Provides: bundled(...)` statements.

> Manual maintenance of bundled provides is cumbersome and error-prone, and given the increasing use of rust in userspace, is not sustainable.

Quite a few packages in Fedora and EPEL also bundle dependencies; by my count, 186 in epel7, 193 in epel8, 215 in epel9, and 790 in rawhide.  I am skeptical of the claim that maintaining bundled provides in RHEL is unsustainable, given that Fedora maintainers (who are mostly volunteers) are able to do it effectively with package versions that change far more often than they do in RHEL.

> The Go SIG recently added means by which bundled provides are automatically generated via %license vendor/modules.txt (that file always existing as a result of go mod vendor).

The problem with that method is that it only handles the bundled provides.  It doesn't ensure the license field accurately reflects all statically linked components, which is the other really important thing that needs to happen when bundling.  At least for the golang packages I maintain in Fedora/EPEL, I still have to examine the bundled components with every version change to validate the license field.  To do that I regenerate the list of bundled provides each time.  @decathorpe has a nice script for doing that for golang packages [here](https://src.fedoraproject.org/rpms/syncthing/blob/rawhide/f/vendor2provides.py).  Hiding this behind a generator could easily lead to the introduction of new bundled components, possibly with non-free licenses.

**Comment by yselkowitz** [comment]
> For the RHEL packages that do not, how does Product Security flag these packages for CVEs in the bundled components? My understanding (which may be flawed or outdated) was that process depends on accurate `Provides: bundled(...)` statements.

Good question, I suspect there may be some gaps that need to be filled.
 
> Quite a few packages in Fedora and EPEL also bundle dependencies; by my count, 186 in epel7, 193 in epel8, 215 in epel9, and 790 in rawhide. I am skeptical of the claim that maintaining bundled provides in RHEL is unsustainable, given that Fedora maintainers (who are mostly volunteers) are able to do it effectively with package versions that change far more often than they do in RHEL.

Most packages that aren't golang or rust perhaps bundle one or a few libraries, whose versions (if there is even) changes infrequently.  Not so with golang or rust, where any given "end" package based on either of these could end up with dozens or even hundreds of bundled dependencies when vendored.  Keeping both the list and versions current with every build manually, even given the limited number of "end" packages in RHEL/ELN which use these, is bound to fail from what I have seen.
 
> > The Go SIG recently added means by which bundled provides are automatically generated via %license vendor/modules.txt (that file always existing as a result of go mod vendor).
> 
> The problem with that method is that it only handles the bundled provides. It doesn't ensure the license field accurately reflects all statically linked components, which is the other really important thing that needs to happen when bundling. At least for the golang packages I maintain in Fedora/EPEL, I still have to examine the bundled components with every version change to validate the license field. To do that I regenerate the list of bundled provides each time. @decathorpe has a nice script for doing that for golang packages [here](https://src.fedoraproject.org/rpms/syncthing/blob/rawhide/f/vendor2provides.py). Hiding this behind a generator could easily lead to the introduction of new bundled components, possibly with non-free licenses.

That's the other half of this concept.  Fedora's cargo macros already have helpers for this, which should similarly be adapted in rust-toolset.  The question here is if those are enough to also handle (the first step of) the provides, or if they can be extended to do so, or if another mechanism is required.

**Comment by gotmax23** [comment]
> That's the other half of this concept. Fedora's cargo macros already have helpers for this, which should similarly be adapted in rust-toolset.

Right, the cargo macros already have support for licensing, and this is a problem whether or not dependencies are bundled. This ticket is specifically about `bundled(crate(...))` Provides.

**Comment by hroncok** [comment]
What we do in setuptools and pip is this:

 - We have [a script to generate the bundled provides](https://src.fedoraproject.org/rpms/python-rpm-generators/blob/rawhide/f/pythonbundles.py).
 - We use it to [put the bundled provides directly to the spec (to a macro)](https://src.fedoraproject.org/rpms/python-pip/blob/fb213c6f74/f/python-pip.spec#_94).
 - We [verify the list in `%check`](https://src.fedoraproject.org/rpms/python-pip/blob/fb213c6f74/f/python-pip.spec#_316).

That way, whenever we [update the package](https://src.fedoraproject.org/rpms/python-pip/c/fb213c6f7), the difference in bundled provides is visible to whoever does it (and the reviewer) directly in the diff. At the same time, we do not risk outdated provides, because the process is designed to fail the build in such case.

The tricky part is obtaining the list from upstream, as Python has no standard for this. It may happen that e.g. pipenv will add a fifth location where they store this information and we might not notice and add it to [the list in spec](https://src.fedoraproject.org/rpms/pipenv/blob/1eae22069d751973d0fc21f1e5d33dcc3bdd5b32/f/pipenv.spec#_218).

As long as we have a reliable mechanism for obtaining the list, I think this approach is quite a good compromise -- it ensures the data is correct and up-to-date while making changes in the list intentionally easy to review.

Considering the usual rate of rebases in RHEL, the list will likely almost never change. (That is not true for ELN, only for RHEL.)

**Comment by decathorpe** [comment]
> That's the other half of this concept.  Fedora's cargo macros already have helpers for this, which should similarly be adapted in rust-toolset.  The question here is if those are enough to also handle (the first step of) the provides, or if they can be extended to do so, or if another mechanism is required.

It should be relatively straightforward to implement support for a mechanism similar to the one that relies on `%license vendor/modules.txt` for Go packages. I can work on this, since there are also (very few) Rust packages in Fedora which would benefit from this, and the implementation will likely be quite similar to the internals of the license breakdown macro.

However, that would still require the macros to be *actually used* (and imported into rust-toolset).

**Comment by carlwgeorge** [comment]
> Most packages that aren't golang or rust perhaps bundle one or a few libraries, whose versions (if there is even) changes infrequently. Not so with golang or rust, where any given "end" package based on either of these could end up with dozens or even hundreds of bundled dependencies when vendored.

I'm aware.  I maintain caddy, which currently has 140 bundled dependencies in rawhide.  By using a script to generate the bundled provides statements, it doesn't really matter to me how many dependencies are bundled (other than checking to make sure the license statement covers any new bundled deps).  This method has scaled well for me as caddy's bundled deps have grown.

> Keeping both the list and versions current with every build manually, even given the limited number of "end" packages in RHEL/ELN which use these, is bound to fail from what I have seen.

It's not required for every build, only builds where the version changes.  It also doesn't have to be completely manual.  There are already two examples in this issue of relatively simple python scripts used to generate the bundled provides statements.  Running a script when the package version changes is not an onerous task.  And @hroncok's suggestion of validating the list in `%check` seems like a great safeguard to me.  Even with the bundled provides conditionalized for RHEL/ELN, the worst case is a Fedora maintainer updates the version, ignores the bundled provides, and the Rawhide build passes while the ELN build fails.  The ELN maintainers can then update the bundled provides and trigger a rebuild.

Let's also talk about the RHEL failure mode if it isn't validated in `%check`.  A hypothetical package gets an initial set of bundled provides that are accurate for version 1.0.  A maintainer updates to package to version 1.1, doesn't check the bundled deps, and doesn't realize that one of the bundled deps got bumped to a higher version.  Product security can still flag the package for any vulnerabilities in that bundled dep with an incorrect version, or any of the other bundled deps that are accurate.  It's better to have a slightly outdated list of bundled deps than no list at all.

**Comment by decathorpe** [comment]
FYI, I have a prototype for a bundled Provides generator for vendored Rust crates ready, I'll do some testing to verify it works as expected, and push it to Fedora over the next few days. I'm not sure what the process would be of getting something like that into ELN / RHEL, though - the RPM macros for Rust / cargo in RHEL are forks of the Fedora macros that haven't been rebased in ages.

**Comment by yselkowitz** [comment]
RHEL uses https://src.fedoraproject.org/rpms/rust/blob/rawhide/f/macros.rust-toolset, so we'll want to sync any changes there.  Perhaps a copr would make sense for testing on ELN before merging anything there.

**Comment by decathorpe** [comment]
Yes, that's on my TODO list. The macros in rust-toolset are ancient forks of those in rust-packaging, missing several years' worth of improvements. So they could use a refresher anyway.

**Comment by yselkowitz** [comment]
The rust-toolset macros in ELN and RHEL 10 now handle this.
